cyral/resource_cyral_repository_user_account.go

```
@@ -0,0 +1,414 @@
+}
+
+type AuthSchemeHashicorpVault struct {
+	Path                 string `json:"path"`
+	IsDynamicUserAccount bool   `json:"isDynamicUserAccount"`
+}
+
+type AuthSchemeEnvironmentVariable struct {
+	VariableName string `json:"variableName"`
+}
+
+type AuthSchemeKubernetesSecret struct {
+	SecretName string `json:"secretName"`
+	SecretKey  string `json:"secretKey"`
+}
+
+type AuthSchemeGCPSecretManager struct {
+	SecretName string `json:"secretName"`
+}
+
+type ApprovalConfig struct {
+	AutomaticGrant            bool `json:"automaticGrant"`
+	MaxAutomaticGrantDuration int  `json:"maxAutomaticGrantDuration"`
+}
+
+type UserAccountConfig struct {
+	Approval *ApprovalConfig `json:"approvalConfig"`
+}
+
+type UserAccountResource struct {
+	UserAccountID    string             `json:"userAccountID"`
+	Name             string             `json:"name"`
+	AuthDatabaseName string             `json:"authDatabaseName"`
+	AuthScheme       *AuthScheme        `json:"authScheme"`
+	Config           *UserAccountConfig `json:"config"`
+}
+
+type CreateUserAccountResponse struct {
+	UserAccountID string `json:"userAccountID"`
+}
+
+func (resp *CreateUserAccountResponse) WriteToSchema(d *schema.ResourceData) error {
+	d.SetId(resp.UserAccountID)
+	return nil
+}
+
+func (response *UserAccountResource) WriteToSchema(d *schema.ResourceData) error {
+	d.SetId(response.UserAccountID)
+	return nil
+}
+
+// ReadFromSchema is used to translate a .tf file into whatever the
+// UserAccounts API expects.
+func (userAccount *UserAccountResource) ReadFromSchema(d *schema.ResourceData) error {
+	userAccount.Name = d.Get("name").(string)
+
+	authSchemeSet := d.Get("auth_scheme").(*schema.Set).List()
+	if len(authSchemeSet) != 1 {
+		return fmt.Errorf("exactly one auth_scheme attribute is required")
+	}
+
+	authSchemes := authSchemeSet[0].(map[string]interface{})
+
+	for k, v := range authSchemes {
+		authSchemeDetails := v.(*schema.Set).List()
+		if len(authSchemeDetails) == 0 {
+			continue
+		}
```

I feel like this should also return an error, but I honestly don't know what all these casts are doing. Let's see what Victor and Alex say.